### PR TITLE
Use size_t when calculating size of all_points

### PR DESCRIPTION
**Pull request by FantasqueX**
Closes: #26642 

Asan log
```
=================================================================
==41401==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7fc55a02a3fc at pc 0x7fc58e304131 bp 0x7ffd54787b00 sp 0x7ffd54787af8
WRITE of size 4 at 0x7fc55a02a3fc thread T0
    #0 0x7fc58e304130 in cv::QRDetectMulti::checkSets(std::vector<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >, std::allocator<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > > > >&, std::vector<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >, std::allocator<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > > > >&, std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >&) /home/fanta/source/opencv/modules/objdetect/src/qrcode.cpp:3726
    #1 0x7fc58e3054b0 in cv::QRDetectMulti::localization() /home/fanta/source/opencv/modules/objdetect/src/qrcode.cpp:3829
    #2 0x7fc58e308020 in cv::ImplContour::detectMulti(cv::_InputArray const&, cv::_OutputArray const&) const /home/fanta/source/opencv/modules/objdetect/src/qrcode.cpp:3987
    #3 0x7fc58e30b5b1 in cv::ImplContour::detectAndDecodeMulti(cv::_InputArray const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, cv::_OutputArray const&, cv::_OutputArray const&) const /home/fanta/source/opencv/modules/objdetect/src/qrcode.cpp:4176
    #4 0x7fc58e28922f in cv::GraphicalCodeDetector::detectAndDecodeMulti(cv::_InputArray const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, cv::_OutputArray const&, cv::_OutputArray const&) const /home/fanta/source/opencv/modules/objdetect/src/graphical_code_detector.cpp:42
    #5 0x5954e8 in Body /home/fanta/source/opencv/modules/objdetect/test/test_qrcode.cpp:48
    #6 0x594fc0 in TestBody /home/fanta/source/opencv/modules/objdetect/test/test_qrcode.cpp:42
    #7 0x67ee6a in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:3919
    #8 0x6734a4 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:3955
    #9 0x641fe8 in testing::Test::Run() /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:3993
    #10 0x6431ac in testing::TestInfo::Run() /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:4169
    #11 0x643d15 in testing::TestCase::Run() /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:4287
    #12 0x659ff3 in testing::internal::UnitTestImpl::RunAllTests() /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:6662
    #13 0x681205 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:3919
    #14 0x675127 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:3955
    #15 0x65734c in testing::UnitTest::Run() /home/fanta/source/opencv/modules/ts/src/ts_gtest.cpp:6271
    #16 0x5907f0 in RUN_ALL_TESTS() /home/fanta/source/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:22240
    #17 0x590cdd in main (/home/fanta/source/opencv-build-4.x-clang/bin/opencv_test_objdetect+0x590cdd) (BuildId: a9363fc788d57c48225fc0559ac9199d07d415db)
    #18 0x7fc58ab242ad in __libc_start_call_main (/lib64/libc.so.6+0x2a2ad) (BuildId: 03f1631dc9760d3e30311fe62e15cc4baaa89db7)
    #19 0x7fc58ab24378 in __libc_start_main@@GLIBC_2.34 (/lib64/libc.so.6+0x2a378) (BuildId: 03f1631dc9760d3e30311fe62e15cc4baaa89db7)
    #20 0x417014 in _start ../sysdeps/x86_64/start.S:115

0x7fc55a02a3fc is located 0 bytes after 2938510332-byte region [0x7fc4aadc8800,0x7fc55a02a3fc)
allocated by thread T0 here:
    #0 0x7fc58e590298 in operator new(unsigned long) (/lib64/libasan.so.8+0xfd298) (BuildId: da72ee674d801ced58193987786b90646d94ff8d)
    #1 0x7fc58e34d010 in std::__new_allocator<cv::Vec<int, 3> >::allocate(unsigned long, void const*) /usr/include/c++/14/bits/new_allocator.h:151

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/fanta/source/opencv/modules/objdetect/src/qrcode.cpp:3726 in cv::QRDetectMulti::checkSets(std::vector<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >, std::allocator<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > > > >&, std::vector<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >, std::allocator<std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > > > >&, std::vector<cv::Point_<float>, std::allocator<cv::Point_<float> > >&)
Shadow bytes around the buggy address:
  0x7fc55a02a100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fc55a02a180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fc55a02a200: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fc55a02a280: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fc55a02a300: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x7fc55a02a380: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00[04]
  0x7fc55a02a400: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7fc55a02a480: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7fc55a02a500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7fc55a02a580: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7fc55a02a600: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==41401==ABORTING
```

`(true_points_group[i].size()` is 1794 and `(true_points_group[i].size() - 2 ) * (true_points_group[i].size() - 1) * true_points_group[i].size())` is 5764222464 which overflows `int`

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by FantasqueX** [comment]
We cannot always use `size_t` because following functions still use `int`. However, we can fix this specific issue by converting to `int` after `/ 6`. Maybe raising some exception when overflow occurs is a better approach?

**Comment by asmorkalov** [comment]
Yes, Assert is very good idea here.

**Comment by asmorkalov** [comment]
I propose to add  true_points_group[i].size() check before computation. Also size_t may have different size.

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -3709,7 +3708,15 @@ bool QRDetectMulti::checkSets(vector<vector<Point2f> >& true_points_group, vecto
...
     {
-        set_size[i] = int( (true_points_group[i].size() - 2 ) * (true_points_group[i].size() - 1) * true_points_group[i].size()) / 6;
+        const std::uint64_t true_points_group_size = true_points_group[i].size();
+        CV_Assert(true_points_group_size - 2 <=
+                  std::numeric_limits<std::uint64_t>::max() / (true_points_group_size - 1));
+        CV_Assert((true_points_group_size - 2) * (true_points_group_size - 1) <=
+                  std::numeric_limits<std::uint64_t>::max() / true_points_group_size);
+        const std::uint64_t combination = (true_points_group_size - 2) *
+                                          (true_points_group_size - 1) * true_points_group_size / 6;
+        CV_Assert(combination <= std::numeric_limits<int>::max());
```
Comment (asmorkalov) [inline modules/objdetect/src/qrcode.cpp:3718]: 3 asserts are overkill here. We have expression like (x-2)*(x-1)*x / 6. It means that single check `CV_Assert(x < limit)` or `CV_Assert(expr < std::numeric_limits<int>::max())` is enough.

**Comment by JSpencerPittman** [comment]
I'm unfamiliar with adding a commit to a pre-existing fork/PR, so I'll just leave this here.

It appears that the main reason for the long duration is the number of `attempts` passed into the kmeans algorithm in the `QRDetectMulti::findNumberLocalizationPoints` and the `QRDetectMulti::findQRCodeContours`. `attempts` is the number of different runs the algorithm does with different initializations (https://docs.opencv.org/4.x/d1/d5c/tutorial_py_kmeans_opencv.html). In the `findNumberLocalizationPoints` the algorithm uses the number of points as the number of attempts. Likewise in `findQRCodeContours` the number of attempts is equal to the number of points across all contours.

A quick summary of the approach is this. `findNumberLocalizationPoints` tries to find (x,y) coordinates in the image that may correspond to one of the corners of a QRCode that has a 1:1:3:1:1 ratio. It will group these together if they're close enough. It tries in multiple passes with varying thresholds. In images with clear QR codes a lesser threshold is needed and it will be found quicker. On images without a qr code present, it will resort to the highest threshold which will include a lot more potential points (2080 potential points in one of the images provided). Points are then grouped together using euclidean distance. Finally the center of the clusters is found using the `cv::kmeans` algorithm. Since there are so many potential points the algorithm runs many more trials (2080 in the image provided).

The `findQRCodeContours` algorithm finds the contours for the image and then groups the points of each contour into a cluster using kmeans that runs for the number of points across all contours.

Maybe something like `std::min(num_points, 10)` and `std::min(num_contours, 10)` or some other constant besides 10 would be a good fix to provide an upper bound on the number of attempts.

**Review by asmorkalov — APPROVED** [review]

**Comment by JSpencerPittman** [comment]
Howdy @marcreichman-pfi, I was referring to [this line](https://github.com/opencv/opencv/blob/4efd52f6763a70e2417795dbbff5fe5613f8d7d0/modules/objdetect/src/qrcode.cpp#L3624), right below it. I believe kmeans assigns a higher priority to the directly passed in number of attempts over the one specified in the termination critiera, `TermCriteria`.
```c++
  kmeans(all_contours_points, count_contours, qrcode_labels,
          TermCriteria( TermCriteria::EPS + TermCriteria::COUNT, 10, 0.1),
          std::min(count_contours, 10), KMEANS_PP_CENTERS, clustered_localization_points);
```

Although, the main choking point is on this line [this line](https://github.com/opencv/opencv/blob/4efd52f6763a70e2417795dbbff5fe5613f8d7d0/modules/objdetect/src/qrcode.cpp#L3567) with a possible fix being:
```c++
    kmeans(list_lines_y, num_points, labels,
            TermCriteria( TermCriteria::EPS + TermCriteria::COUNT, 10, 0.1),
            std::min(num_points, 10), KMEANS_PP_CENTERS, tmp_localization_points);
```

This alteration sped it up on my end to almost instant rather than the 10 minutes of waiting. These changes may cause an error, but I believe it's a further downstream issue that needs to be fixed. I've been getting a `std::bad_alloc`.

### Performance Comparison
I went ahead and wrote a custom benchmark using 1000 images from this [dataset](https://universe.roboflow.com/qrdetection-eohio/qrcodedetection-sovr6). On it the results before and after the change were identical on images that had a QR code within them. The code was quite confusing, but it seems like when it finds a prominent tag it quits its search and moves on with it. So often the problem is it has too few contours which is why its search is quick, despite leaving out other QR codes. However, when there's no QR code it's grabbing onto anything it can leaving a pile of 1000 contours too sift through on a singular image.

img_with_dets

*Here the green boxes are the ground truth and the red square is the detected QR code.*

### Results
Against the $1000$ images which possessed $1321$ annotations in total, I counted how many qr codes were detected, how many went undetected, and how many detection were false. The dataset provides ground truth bounding boxes. I considered two bounding boxes to be equivalent if there IoU (intersection over union) was greater than $0.8$. 

| Method | Matches | Undetected | False Detections |
| --- | --- | --- | --- |
| Original | 426 | 895 | 48 |
| Modified | 426 | 895 | 48 |
| ArUco | 645 | 676 | 163 |

I went ahead and also benchmarked the ArUco detector to see how it compares.

**Pull request merged**